pkg/controller/appgroup.go

```diff
@@ -0,0 +1,423 @@
+		AddFunc:    ctrl.agAdded,
+		UpdateFunc: ctrl.agUpdated,
+		DeleteFunc: ctrl.agDeleted,
+	})
+
+	klog.V(5).InfoS("Setting up Pod event handlers")
+	podInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    ctrl.podAdded,
+		UpdateFunc: ctrl.podUpdated,
+		DeleteFunc: ctrl.podDeleted,
+	})
+
+	ctrl.agLister = agInformer.Lister()
+	ctrl.podLister = podInformer.Lister()
+	ctrl.agListerSynced = agInformer.Informer().HasSynced
+	ctrl.podListerSynced = podInformer.Informer().HasSynced
+	ctrl.agClient = agClient
+	return ctrl
+}
+
+// Run : starts listening on channel events
+func (ctrl *AppGroupController) Run(workers int, stopCh <-chan struct{}) {
+	defer ctrl.agQueue.ShutDown()
+
+	klog.InfoS("Starting App Group controller")
+	defer klog.InfoS("Shutting App Group controller")
+
+	if !cache.WaitForCacheSync(stopCh, ctrl.agListerSynced, ctrl.podListerSynced) {
+		klog.Error("Cannot sync caches")
+		return
+	}
+	klog.InfoS("App Group sync finished")
+	for i := 0; i < workers; i++ {
+		go wait.Until(ctrl.worker, time.Second, stopCh)
+	}
+
+	<-stopCh
+}
+
+// agAdded : reacts to a AppGroup creation
+func (ctrl *AppGroupController) agAdded(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	ag := obj.(*v1alpha1.AppGroup)
+
+	// If startScheduleTime - createTime > 2days, do not enqueue again because pod may have been GCed
+	if ag.Status.RunningWorkloads == 0 &&
+		ag.Status.ScheduleStartTime.Sub(ag.CreationTimestamp.Time) > timeLimitation {
+		return
+	}
+
+	klog.V(5).InfoS("Enqueue AppGroup ", "app group", key)
+	ctrl.agQueue.Add(key)
+}
+
+// agUpdated : reacts to a AppGroup update
+func (ctrl *AppGroupController) agUpdated(old, new interface{}) {
+	ctrl.agAdded(new)
+}
+
+// agDeleted : reacts to a AppGroup deletion
+func (ctrl *AppGroupController) agDeleted(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	klog.V(5).InfoS("Enqueue deleted app group key", "appGroup", key)
+	ctrl.agQueue.AddRateLimited(key)
+}
+
+// podAdded : reacts to a pod creation
+func (ctrl *AppGroupController) podAdded(obj interface{}) {
+	pod := obj.(*v1.Pod)
+	agName := util.GetPodAppGroupLabel(pod)
+	if len(agName) == 0 {
+		return
+	}
+	ag, err := ctrl.agLister.AppGroups(pod.Namespace).Get(agName)
+	if err != nil {
+		klog.ErrorS(err, "Error while adding pod")
+		return
+	}
+	klog.V(5).InfoS("Add App group when pod gets added", "AppGroup", klog.KObj(ag), "pod", klog.KObj(pod))
+	ctrl.agAdded(ag)
+}
+
+// podDeleted : reacts to a pod deletion
+func (ctrl *AppGroupController) podDeleted(obj interface{}) {
+	_, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	ctrl.podAdded(obj)
+}
+
+// pgUpdated : reacts to a pod update
+func (ctrl *AppGroupController) podUpdated(old, new interface{}) {
+	ctrl.podAdded(new)
+}
```

Comment:
In the previous design, we were updating the CR via the controller when pods were rescheduled to other nodes. In the current design, perhaps it does not make sense to sync the AppGroup when pods are updated. So, it might be indeed removed to avoid multiples sync without a clear purpose...